### PR TITLE
feat: add alchemyTokenMethods

### PR DESCRIPTION
**Pull request by erik-rt**
Time spent on this PR: 4 hours

Resolves: #331 

# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing

# What is the new behavior?

This adds two methods, `token_allowance` and `token_metadata` which implement Alchemy's similarly named methods. I also add a model for metadata and a model for allowance.

# Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: crates/eth-rpc/src/servers/alchemy_rpc.rs
Comment (Eikix) [inline crates/eth-rpc/src/servers/alchemy_rpc.rs:41]: Could this work?
`.map_err(EthApiError::<P::Error>::from)`

File: crates/eth-rpc/src/servers/alchemy_rpc.rs
Comment (Eikix) [inline crates/eth-rpc/src/servers/alchemy_rpc.rs:61]: can we remove this unwrap?

File: crates/eth-rpc/src/servers/alchemy_rpc.rs
Comment (Eikix) [inline crates/eth-rpc/src/servers/alchemy_rpc.rs:39]: I wonder in general if we should cast 256-bit keccak into a 252 felt, with the possibility that the conversion may fail.
It'd be safer to pack it into two felts in case of failure, to not damage business logic, 
The packed two felts would then be inlined or destructured into calldata anyway. I may be wrong. 

This is a general question about the flow of the RPC and is pretty important, @ClementWalter may have an opinion

File: crates/eth-rpc/src/servers/alchemy_rpc.rs
Comment (Eikix) [reply crates/eth-rpc/src/servers/alchemy_rpc.rs:78]: In case it doesn't work, we should probably open an issue to implement error conversion for these two error types, in order to minimize the use of map_err in our codebase.

File: crates/eth-rpc/src/servers/alchemy_rpc.rs
Comment (Eikix) [inline crates/eth-rpc/src/servers/alchemy_rpc.rs:95]: is this truly a handle? it is awaited

Comment (Eikix) [reply crates/eth-rpc/src/servers/alchemy_rpc.rs:95]: you could technically parallelize these handles by creating three futures (remove the .await on the handle)
and then use futures crate join_all to await them.
it is used in core>client>mod>token_balances

File: crates/eth-rpc/src/servers/alchemy_rpc.rs
Comment (Eikix) [inline crates/eth-rpc/src/servers/alchemy_rpc.rs:126]: try to remove this unwrap() from the flow

File: crates/core/src/models/metadata.rs
Comment (Eikix) [inline crates/core/src/models/metadata.rs:7]: Should these fields be String or &str? 
It's an open question, I'm not yet aware of the trade-offs associated

**Review by Eikix — COMMENTED** [review]
Awesome first PR! Thank you for submitting it!

Did you join our discord yet? 
discord.gg/kakarotzkevm 
We can chat there further if need be in the devs chat

Nice first PR!
You got the hang of it quite well,
Could you add testing flow in your PR? no need to add tests in eth-rpc for now, I'm refactoring tests there, but if you could use the flow from 
core>tests>client.rs like `test_counter`
replace deployment of counter by an erc20 and try your new routes

**Comment by erik-rt** [comment]
> Did you join our discord yet? discord.gg/kakarotzkevm We can chat there further if need be in the devs chat

Yes my username in there is etash. I actually contributed during the early days of the project under @etashhh but I got busy with work and now that I have more time I wanted to consolidate my GitHub activity under this account haha

> Nice first PR! You got the hang of it quite well, Could you add testing flow in your PR? no need to add tests in eth-rpc for now, I'm refactoring tests there, but if you could use the flow from core>tests>client.rs like `test_counter` replace deployment of counter by an erc20 and try your new routes

Sure thing, I'll get to it 🤝

File: crates/eth-rpc/src/servers/alchemy_rpc.rs
Comment (erik-rt) [reply crates/eth-rpc/src/servers/alchemy_rpc.rs:41]: That works 👍

File: crates/core/src/models/metadata.rs
Comment (erik-rt) [reply crates/core/src/models/metadata.rs:7]: My understanding is that if we don't know the contents of the string at runtime, it makes more sense to use `String`. Also using `&str` would involve setting up unnecessary lifetimes in this case.

File: crates/eth-rpc/src/servers/alchemy_rpc.rs
Comment (erik-rt) [reply crates/eth-rpc/src/servers/alchemy_rpc.rs:39]: I'll need guidance on this one, I'm not familiar with the tradeoffs between the two choices. I followed the `token_balances()` logic in the `core/src/client/mod.rs` file for this.

**Comment by Eikix** [comment]
> > Did you join our discord yet? discord.gg/kakarotzkevm We can chat there further if need be in the devs chat
> 
> Yes my username in there is etash. I actually contributed during the early days of the project under @etashhh but I got busy with work and now that I have more time I wanted to consolidate my GitHub activity under this account haha
> 
> > Nice first PR! You got the hang of it quite well, Could you add testing flow in your PR? no need to add tests in eth-rpc for now, I'm refactoring tests there, but if you could use the flow from core>tests>client.rs like `test_counter` replace deployment of counter by an erc20 and try your new routes
> 
> Sure thing, I'll get to it 🤝

Nice to have you back, etash! I remember! You mentioned the Rust RPC in the earliest days
PS you might need a rebase, there were some bugfixes in the most recent merges

File: crates/eth-rpc/src/servers/alchemy_rpc.rs
Comment (Eikix) [reply crates/eth-rpc/src/servers/alchemy_rpc.rs:39]: I think this fixes it:
`FieldElement::from_byte_slice_be(&keccak256("balanceOf(address)").0[0..4])?;`

Comment (Eikix) [reply crates/eth-rpc/src/servers/alchemy_rpc.rs:39]: Granted you implement 
```rust
impl<E: std::error::Error> From<FromByteSliceError> for EthApiError<E> {
    fn from(err: FromByteSliceError) -> Self {
        Self::ConversionError(format!("Failed to convert from byte slice: {}", err))
    }
}
```

Comment (erik-rt) [reply crates/eth-rpc/src/servers/alchemy_rpc.rs:39]: Are you running into the same error in that `impl doesn't use only types from inside the current crate`? I'm seeing an issue in that `FromByteSliceError` and `EthApiError` aren't defined in the current crate.

File: crates/core/src/models/metadata.rs
Comment (Eikix) [reply crates/core/src/models/metadata.rs:7]: Works for me!

File: crates/eth-rpc/src/servers/alchemy_rpc.rs
Comment (Eikix) [reply crates/eth-rpc/src/servers/alchemy_rpc.rs:39]: Ahh sure! You'd have to put the above code snippet inside in crates>core>src>client>errors.rs

Comment (erik-rt) [reply crates/eth-rpc/src/servers/alchemy_rpc.rs:39]: That makes sense haha

Comment (erik-rt) [reply crates/eth-rpc/src/servers/alchemy_rpc.rs:39]: So the above code works in `token_balances` implementation in `mod.rs` but in the context of `alchemy_rpc.rs` I'm seeing that it is necessary to add `.map_err(EthApiError::<P::Error>::from)?;`

Comment (Eikix) [reply crates/eth-rpc/src/servers/alchemy_rpc.rs:39]: Do we know why? Is it a problem of modules/scope in the sense that the conversion is not properly understood?

File: crates/eth-rpc/src/servers/alchemy_rpc.rs
Comment (Eikix) [reply crates/eth-rpc/src/servers/alchemy_rpc.rs:61]: Could we try with a "?" question mark
And then try to handle the error conversion as best we can (map_err being last resort)

File: crates/eth-rpc/src/servers/alchemy_rpc.rs
Comment (erik-rt) [reply crates/eth-rpc/src/servers/alchemy_rpc.rs:39]: Exactly, if I try to use the `?` operator without first using `map_err` then I'm met with the following error:

```rust
error[E0277]: `?` couldn't convert the error to `ErrorObject<'_>`
  --> crates/eth-rpc/src/servers/alchemy_rpc.rs:40:101
   |
40 |         let entry_point = FieldElement::from_byte_slice_be(&keccak256("balanceOf(address)").0[0..4])?;
   |                                                                                                     ^ the trait `From<FromByteSliceError>` is not implemented for `ErrorObject<'_>`
   |
   = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
   = help: the following other types implement trait `From<T>`:
             <ErrorObject<'a> as From<jsonrpsee::jsonrpsee_types::error::ErrorCode>>
             <ErrorObject<'static> as From<EthApiError<E>>>
             <ErrorObject<'static> as From<ForkchoiceUpdateError>>
   = note: required for `std::result::Result<TokenAllowance, ErrorObject<'_>>` to implement `FromResidual<std::result::Result<std::convert::Infallible, FromByteSliceError>>`
```

File: crates/eth-rpc/src/servers/alchemy_rpc.rs
Comment (erik-rt) [reply crates/eth-rpc/src/servers/alchemy_rpc.rs:61]: I managed to make the following work:

```rust
let allowance = U256::try_from_be_slice(call.as_ref())
    .ok_or(EthApiError::<P::Error>::ConversionError("error converting from Bytes to U256".into()))?;
```

**Comment by erik-rt** [comment]
Hey @Eikix I believe I'm almost done. The final task is to upload tests. I added boilerplate to the `core/tests/client.rs` file under `test_alchemy_api` but I'm wondering if you can provide any guidance on:

1. How we should be mapping the bytes from an Ethereum call response to a `&str` in the context of asserting equality of the name and symbol of the token
2. How to set up the Alchemy RPC in the context of this test in order to test routes.

**Comment by Eikix** [comment]
> Hey @Eikix I believe I'm almost done. The final task is to upload tests. I added boilerplate to the `core/tests/client.rs` file under `test_alchemy_api` but I'm wondering if you can provide any guidance on:
> 
> 1. How we should be mapping the bytes from an Ethereum call response to a `&str` in the context of asserting equality of the name and symbol of the token
> 2. How to set up the Alchemy RPC in the context of this test in order to test routes.

Hey Erik! Thanks for your patience,
You're right, the PR is almost done & ready

One thing I noticed is that the test you added is in `core` crate, whereas the alchemy API can only be accessed through eth-rpc crate.
The current structure of the repository is that `core` crate packs the kakarot_client, and `eth-rpc` wraps the kakarot_client with eth rpc route. That is to say `eth-rpc` crate is outward facing, and depends on `core` crate.
Meaning that you'll need the flow of #365 to add tests to your alchemy api route inside eth-rpc

About question 1. I'm not sure
For question 2. I mentioned the solution above -> this hasn't been done yet, but essentially we need to set up integration tests in eth-rpc -> we need to be able to write the tests from within eth-rpc crate.

Side note: An alternative to the problem of being able to test your logic from within `core` crate would be to put most of the logic inside kakarot_client and call it from eth-rpc (akin to most routes).

**Comment by erik-rt** [comment]
Okay sounds good, it looks like PR #370 will solve a lot of the setup so I'll plan to wait until that's merged into `main` to finish up the Alchemy API testing integration

**Comment by ClementWalter** [comment]
I guess that this is ready for work again @ftupas @greged93 can you give more details about the tech strategy to follow?

**Comment by greged93** [comment]
I think the best would be to look at the rework that was done for the TokenBalance endpoint, allowance and metadata should pretty much be the same. Also I quickly glanced over the PR and I saw a lot of code in the eth-rpc, but again @erik-rt you can look at how we do it for token_balances, eth-rpc should just forward the call to the core crate.

**Comment by erik-rt** [comment]
Hey @greged93 that sounds good, I'll start working on it again. I want to verify that we shouldn't be writing the code in eth-rpc as in issue #331 it was recommended that we write the methods inline rather than how token_balances is set up.

**Comment by greged93** [comment]
Yes indeed checked the issue, so you should define the server as you did in eth-rpc but the implementation should be in core, the eth-rpc function should just forward the call, I can confirm :)

**Comment by erik-rt** [comment]
Just wanted to update that I'm still working on this, just got swamped with other work this week 😁

**Comment by greged93** [comment]
Just ping me whenever we can review :)

**Comment by erik-rt** [comment]
Hey @greged93 I finally have time again 😁 I believe I've finished the `token_allowance` endpoint but wanted your guidance on how to proceed with `token_metadata`:

1. In `models/metadata.rs` am I defining the futures correctly, using `F` for a `U256` result and `G` for a `String` result? 
https://github.com/erik-rt/kakarot-rpc/blob/488379fa8a117d88cd4e2058302f6525fe337423/crates/core/src/models/metadata.rs#L20-L32

2. In `contracts/erc20/ethereum_erc20.rs` should I be splitting out the functions as I have them currently?
https://github.com/erik-rt/kakarot-rpc/blob/488379fa8a117d88cd4e2058302f6525fe337423/crates/core/src/contracts/erc20/ethereum_erc20.rs#L79-L95

3. In `client/mod.rs` do you have any guidance on assembling the handles assuming the code above stays the same?
https://github.com/erik-rt/kakarot-rpc/blob/488379fa8a117d88cd4e2058302f6525fe337423/crates/core/src/client/mod.rs#L449-L457

File: crates/core/src/contracts/erc20/ethereum_erc20.rs
Comment (Eikix) [inline crates/core/src/contracts/erc20/ethereum_erc20.rs:92]: isn't name a string in solidity? it could be of arbitrary length, right?
i tried locally on remix and you can put very long names

File: crates/core/src/contracts/erc20/ethereum_erc20.rs
Comment (Eikix) [inline crates/core/src/contracts/erc20/ethereum_erc20.rs:110]: same for symbol, string is of arbitrary length

File: crates/core/src/contracts/erc20/ethereum_erc20.rs
Comment (erik-rt) [reply crates/core/src/contracts/erc20/ethereum_erc20.rs:92]: Oh yeah I had those as placeholders for the time being to get things to compile; looking at it again should I just use `SignatureDecodingError` or `TransactionDecodingError` instead?

Comment (ClementWalter) [reply crates/core/src/contracts/erc20/ethereum_erc20.rs:92]: wdyt about adding another variant, like `DataDecodingError::InvalidBytesString` ?

File: crates/core/src/contracts/erc20/ethereum_erc20.rs
Comment (ClementWalter) [reply crates/core/src/contracts/erc20/ethereum_erc20.rs:110]: see above, proposition to add another variant

File: crates/core/src/contracts/erc20/ethereum_erc20.rs
Comment (ClementWalter) [inline crates/core/src/contracts/erc20/ethereum_erc20.rs:126]: see above, proposition to add another variant

File: crates/eth-rpc/src/servers/alchemy_rpc.rs
Comment (ClementWalter) [inline crates/eth-rpc/src/servers/alchemy_rpc.rs:49]: should use kakarot_client here right?

File: crates/core/src/client/api.rs
Comment (greged93) [inline crates/core/src/client/api.rs:59]: `token_allowance`, `token_balances` and `token_metadata` don't actually belong to the Eth API, what do you think of making a new trait for Alchemy methods?

File: crates/core/src/client/mod.rs
Comment (greged93) [inline crates/core/src/client/mod.rs:393]: This is a very strange syntax, why not just:
```suggestion
        contract.allowance(
            account_address.into(),
            spender_address.into(),
            block_id,
        ).await
```

File: crates/core/src/models/allowance.rs
Comment (greged93) [inline crates/core/src/models/allowance.rs:1]: I think there as been a bit of a misunderstanding as to why we implemented the `TokenBalances` in the way we did, using `FutureTokenBalance` and implementing `Future for FutureTokenBalance`: the main reason was that we needed a way to return futures in a loop, which could then be joined with `join_all` and resolved asynchronously, all the while linking the balance we got to the address of the contract we got this balance from. You can see all this [here](https://github.com/kkrt-labs/kakarot-rpc/blob/main/crates/core/src/client/mod.rs#L389). However, in your case, you are just calling `token_allowance`, retrieving the allowance for one owner, one spender and one contract. There is no need to implement all the future resolution handling.

Please let me know if this is clear and otherwise DM me on telegram.

File: crates/core/src/models/metadata.rs
Comment (greged93) [inline crates/core/src/models/metadata.rs:1]: Same comment as above, no need to implement all of this.

File: crates/eth-rpc/src/servers/alchemy_rpc.rs
Comment (greged93) [reply crates/eth-rpc/src/servers/alchemy_rpc.rs:49]: yes call should be forwarded to kakarot_client

**Review by greged93 — CHANGES_REQUESTED** [review]
Couple of comments, I think there as been a misunderstanding as to why we implement a `FutureTokenBalance`, which doesn't seem needed here

File: crates/core/src/client/api.rs
Comment (erik-rt) [reply crates/core/src/client/api.rs]: Do you have any preferences? The logic is almost the same as in the `U256` case so I'm wondering how different the new error variant should be.

File: crates/core/src/contracts/erc20/ethereum_erc20.rs
Comment (erik-rt) [reply crates/core/src/contracts/erc20/ethereum_erc20.rs:124]: Do you have any preferences? The logic is almost the same as in the U256 case so I'm wondering how different the new error variant should be.

**Comment by greged93** [comment]
Will review! I guess it makes sense to have the AlchemyAPI be another PR, can you just open an issue for it?

File: crates/core/src/client/api.rs
Comment (erik-rt) [reply crates/core/src/client/api.rs]: I initially started moving over the methods to a new trait implementation but this involved updating how the RpcModule is built at the top and I was getting into some messy logic. Specifically, in `eth-rpc/src/rpc.rs` we would have to add another argument to the `new` method that takes a `KakarotAlchemyApi` type and pass that into the `AlchemyRpc::new()` call. Not sure how you'd want this handled.

**Comment by Eikix** [comment]
Hey @erik-rt, thanks a lot for being so patient on this PR, it has definitely been a long and hard journey. Thanks for bearing with us.

As we went, we've uncovered some tech debt and hard stuff to account for,
This is partly because merging this PR was hard to begin with.

**Comment by Eikix** [comment]
Hi @erik-rt, sorry again for the delay, disorganization and pain throughout this long PR,
We've gone back and forth on where the trait / impl should live, how to implement it, etc.

I think one main problem is our current architecture.
I'm looking at a way to refactor it to make it sounder, and less redundant.

I'm closing this PR. I'd like to thank you for being so patient and apologize for the time lost. When we have a leaner architecture, it'll be way easier to reimplement this PR, and I hope you'll want to give it another shot.

Thanks

**Comment by erik-rt** [comment]
Yeah sounds good, no worries I understand. Best of luck going forward.

**Pull request closed without merging**
